### ISSUE TITLE
I don’t see ely.by skins, even with authlib injector.

### ISSUE THREAD

**Issue by 1Grobotron1** [unmojang/FjordLauncher#44]
### Operating System

Windows

### Version of Fjord Launcher

Fjord Launcher 8.3.

### Version of Qt

Qt 6.6.3.

### Description of bug

Skins from authlib injector from ely.by do not work, in multiplayer I don’t see my own or someone else’s skin, in singleplayer everything works.

### Steps to reproduce

Connect your ely.by account to authlib injector, and connect to any server.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

**Comment by evan-goode** [comment]
Are you playing on `online-mode=false` servers, or are the Minecraft servers configured to use Ely.by's authentication servers with `online-mode=true`? If it's the former, this is probably https://github.com/unmojang/FjordLauncher/issues/18.

**Comment by 1Grobotron1** [comment]
> Are you playing on `online-mode=false` servers, or are the Minecraft servers configured to use Ely.by's authentication servers with `online-mode=true`? If it's the former, this is probably #18.
#18 is not solved.

**Comment by evan-goode** [comment]
So is this a duplicate of #18? #18 is unlikely to be resolved by Fjord Launcher, authlib-injector, or Ely.by. Showing skins on `online-mode=false` servers is a bit outside the scope of these projects. Try using the Forge/Fabric mod https://github.com/xfl03/MCCustomSkinLoader. I think it will load skins from Ely.by (among other sources) by default.

**Comment by 1Grobotron1** [comment]
> So is this a duplicate of #18? #18 is unlikely to be resolved by Fjord Launcher, authlib-injector, or Ely.by. Showing skins on `online-mode=false` servers is a bit outside the scope of these projects. Try using the Forge/Fabric mod https://github.com/xfl03/MCCustomSkinLoader. I think it will load skins from Ely.by (among other sources) by default.

But I have another problem, others see my skin when they are with ely.by.
Do I also need to download MCCustomSkinLoader on the server, and download it to all my friends?

**Comment by evan-goode** [comment]
The reason they're seeing skins is maybe because they're using [Ely.by's patched authlib](https://ely.by/load). Fjord Launcher doesn't use this since it's proprietary and Ely.by-specific, and we want to work with all authentication servers. Ely.by's patched authlib loads skins by player name in offline mode, similar to CustomSkinLoader.

CustomSkinLoader is client-side only, so yes, if you want other Fjord Launcher users to see each other's skins on `online-mode=false` servers, you'd need each user to install CustomSkinLoader. You could also look into https://github.com/samolego/FabricTailor which is a server-side mod that works with Ely.by.

But if everyone playing on the server is already using Ely.by, then I would set `online-mode=true` on the server and configure the server to authenticate users with Ely.by. That's the recommended way of doing things.

**Comment by yum13241** [comment]
I've always been a proponent of letting users use Ely.by's patched authlib **if they want**, because while it's Ely.by specific, there are the following reasons this should be an option:

1. [ElyPrismLauncher](https://github.com/ElyPrismLauncher/ElyPrismLauncher/) is a thing and does this. UltimMC used to do this, but uses authlib-injector now (making you look like Steve on your own client).
2. It emulates TL Legacy's (read: the real, safe TLauncher) behavior, and that's how Ely.by wants you to use the skins, not by having to add a mod to ALL (CustomSkinLoader) of your instances. It also means other people can see your skin without having to mod their game, since authlib mods are universal.
3. User choice is part of F(L)OSS software development anyway, and this issue reflects that (to some degree).

> But if everyone playing on the server is already using Ely.by, then I would set online-mode=true on the server and configure the server to authenticate users with Ely.by. That's the recommended way of doing things.

You cannot guarantee this, plus skins are seen as "optional" by many players, some just don't care and want to just play on your server. I speak as someone with several years experience of hosting Minecraft servers.

What I'd do is add a dropdown letting you "choose your authlib provider" with the options being "authlib-injector (universal)" or "elyby-authlib (ely.by only)", and choosing which one to automatically download with the existing "automatically install authlib-injector" dropdown.
Of course, there'd be a warning if you try to launch an instance using elyby-authlib when not using an ely.by account, and it would default to authlib-injector, and to save everyone a burden, there'd be per-instance overrides, and the possibility of having a new "authlib provider" retroactively apply to all instances that don't have their specific settings, like how the rest of Prism Launcher and its forks are handled.

**Comment by evan-goode** [comment]
I've created a soft fork of authlib-injector at https://github.com/unmojang/authlib-injector and I added a feature to fetch missing skins+capes by player name, which allows skins to work seamlessly on `online-mode=false` servers. I intend to make builds of the fork available in the "Add authlib-injector" screen alongside the official authlib-injector builds, and mark the fork builds as "recommended" so they are automatically chosen to be installed when launching an instance with an authlib-injector account.

If you'd like to test this now (please do!), you can open Fjord Launcher, go to Settings -> APIs -> Services -> Metadata Server and change the url to https://unmojang.github.io/meta-launcher-test/v1/. Then edit an instance, go to Version, select the authlib-injector entry, and click "Change Version" (or click "Install authlib-injector" if authlib-injector is not yet installed to the instance). Then select the version `1.2.5.1.unmojang`.

**Comment by evan-goode** [comment]
Hmm, I noticed that my patch is incompatible with Fabric. Back to the drawing board.

**Comment by evan-goode** [comment]
Fabric incompatibility should be fixed with `1.2.5.2.unmojang`.

**Comment by yum13241** [comment]
Nice, but I already use ElyPrismLauncher lol

**Comment by catfromplan9** [comment]
> I've always been a proponent of letting users use Ely.by's patched authlib **if they want**, because while it's Ely.by specific, there are the following reasons this should be an option:
> 
>     1. [ElyPrismLauncher](https://github.com/ElyPrismLauncher/ElyPrismLauncher/) is a thing and does this. UltimMC used to do this, but uses authlib-injector now (making you look like Steve on your own client).
> 
>     2. It emulates TL Legacy's (read: the real, safe TLauncher) behavior, and that's how Ely.by wants you to use the skins, not by having to add a mod to ALL (CustomSkinLoader) of your instances. It also means other people can see your skin without having to mod their game, since authlib mods are universal.
> 
>     3. User choice is part of F(L)OSS software development anyway, and this issue reflects that (to some degree).
> 
> 
> > But if everyone playing on the server is already using Ely.by, then I would set online-mode=true on the server and configure the server to authenticate users with Ely.by. That's the recommended way of doing things.
> 
> You cannot guarantee this, plus skins are seen as "optional" by many players, some just don't care and want to just play on your server. I speak as someone with several years experience of hosting Minecraft servers.
> 
> What I'd do is add a dropdown letting you "choose your authlib provider" with the options being "authlib-injector (universal)" or "elyby-authlib (ely.by only)", and choosing which one to automatically download with the existing "automatically install authlib-injector" dropdown. Of course, there'd be a warning if you try to launch an instance using elyby-authlib when not using an ely.by account, and it would default to authlib-injector, and to save everyone a burden, there'd be per-instance overrides, and the possibility of having a new "authlib provider" retroactively apply to all instances that don't have their specific settings, like how the rest of Prism Launcher and its forks are handled.

Re. "authlib mods are universal": elyby is not distributing an "authlib mod", they're distributing modified copies of authlib. This is what people did with Minecraft itself in the stone age, before they invented mod loaders some time around 2010. Unfortunately, it seems the team at ely.by has not gotten the memo yet. Patching authlib by providing modified copies (proprietary, by the way) for every single version is an extremely stupid idea, and it violates copyright law. You can use a different launcher if you would like, but this will not be supported. authlib-injector could be considered an "authlib mod" since it transforms the game code instead of directly forking and restributing it. You are correct when you mentioned that authlib mods are universal in the sense that they do not require releases for each minecraft version.

As for the feature you requested, resolving UUIDs from player names on offline mode servers is not a closely guarded and well-kept secret of elyby's bootleg authlib. The root cause of the issue is that upstream authlib-injector lacks a transformer for it (only Evan's fork has it), but support is included for this in Loki and this will be added to Fjord Launcher soon.

I think we can close this issue and #18 when Loki support lands.

**Comment by yum13241** [comment]
> Re. "authlib mods are universal": elyby is not distributing an "authlib mod", they're distributing modified copies of authlib.

Pedantic much? @Octol1ttle

> Unfortunately, it seems the team at ely.by has not gotten the memo yet. Patching authlib by providing modified copies (proprietary, by the way) for every single version is an extremely stupid idea, and it violates copyright law.

We aren't paid to do M$'s job here. M$ has never taken action against auth servers ever.
`elyby-authlib` resolves skins even in offline mode, `authlib-injector` does not. That's what's causing all the ruckus and confusion anyway. As I'm sure we all know, `authlib` acts differently at some version milestones.
The average ely.by user's not going to patch the game manually, they'll either use "Legacy Launcher" (aka the real, safe TLauncher), or something like PineconeMC (made by the guy I tagged earlier).

**Comment by Octol1ttle** [comment]
> Pedantic much? @Octol1ttle

why am I tagged here? what do I have to do with all of this? arguing about phrasing and semantics is a waste of time

**Comment by catfromplan9** [comment]
> The average ely.by user's not going to patch the game manually,

Yeah, we have had authlib-injector since around 2017 for this exact purpose. Thats what it was invented for.

>  `elyby-authlib` resolves skins even in offline mode, `authlib-injector` does not. 

Please see:

> As for the feature you requested, resolving UUIDs from player names on offline mode servers is not a closely guarded and well-kept secret of elyby's bootleg authlib. The root cause of the issue is that upstream authlib-injector lacks a transformer for it (only Evan's fork has it), but support is included for this in Loki and this will be added to Fjord Launcher soon.

> I think we can close this issue and https://github.com/unmojang/FjordLauncher/issues/18 when Loki support lands.

There is nothing whatsoever the elyby patches offer other than a DMCA risk, wasted development time integrating them, and hardcoded support for 1 (one) API server. Elyby is not special, they are one of many API servers to be supported within the Fjord Launcher.

**Comment by catfromplan9** [comment]
> > Pedantic much? [@Octol1ttle](https://github.com/Octol1ttle)
> 
> why am I tagged here? what do I have to do with all of this? arguing about phrasing and semantics is a waste of time

Feel free to unsubscribe and ignore the issue

**Comment by catfromplan9** [comment]
TLDR I like the idea of fixing the original subject of this issue, I did fix it in Loki, and the fix lands soon when launcher is updated

**Comment by yum13241** [comment]
> > Pedantic much? [@Octol1ttle](https://github.com/Octol1ttle)
> 
> why am I tagged here? what do I have to do with all of this? arguing about phrasing and semantics is a waste of time

you're the developer of ely.by's authlib?
